Tests: Improve testing of `NotifyDBus`, `NotifyGnome`, and `NotifyMacOSX`

amotl opened the pull request:
Dear Chris,

while working on adjusting the test suite with #687, I discovered a few spots where I think the test cases may get improved.  I hope you still like my patches.

The spots are specifically those places where `reload_plugin` needs to be used, in order to examine module-level plugin code, specifically at the `NotifyDBus`, `NotifyGnome`, and `NotifyMacOSX` plugins.

With kind regards,
Andreas.

<sub>
P.S.: When you think the patch will be ready for merging, can I ask you to keep the commits as they are, instead of squashing them? In this manner, individual changes can be reverted much easier, keep their own commit message about the corresponding rationale, and are just so much easier to handle, followup on, or reference, for example when talking about code changes in retrospective.
</sub>



### In a nutshell

By using proper [`pytest` fixtures](https://docs.pytest.org/en/latest/explanation/fixtures.html) for context/environment setup, as well as [`pytest-mock`](https://pypi.org/project/pytest-mock/) for mocking, the patch significantly improves convenience and structure both when writing, and when reading and reasoning about test cases.


### Benefits

Because the test setup code is now available as fixtures, it is easy to use a dedicated function per test case, without the need to duplicate setup code.

It is now guaranteed that each test case gets a fresh and standardized context/environment, without the chance that something is leaked from a previous test, and without the need to manually run infrastructural code in between the test cases, which is both tedious and error-prone. Specifically, when writing test cases in such a manner, there is no need for manual "undo" code after mangling the underlying machinery, so it will never break on this aspect.

### My personal learning path

While I am a big fan of Python decorators in general, I would never want to look back before I learned about `pytest-mock`. It is just so much easier now to refactor and modularize test cases, and to provide setup/teardown code to them.

This is my stake on this topic.

- Manipulating object references back and forth directly: Not encouraged, too much can go wrong and leakage is imminent.
- Mocks with functional interface: No context, no guaranteed cleanup. Not much better than the first option.
- Mocks with decorator interface: Minimal amount of code needed, cleanup guaranteed; great if you really just need to slap it on. Difficult when needing to refactor test cases, as taking things on and off needs swapping out _both_ the decorator, as well as the corresponding function argument, which quickly gets boring and tedious.
- Mocks with context manager interface: More flexibility in setting up customized mocks, cleanup guaranteed. However, when needing to mock a bunch of objects, code with many indentation levels quickly and clearly feels like it gets entangled into Python syntax specialties.
- Mocks with `pytest-mock`: Very clean interface, cleanup guaranteed. Easy to edit, as it is just regular, imperative Python code, without using any idioms like decorators or dependency injection on the function argument level. In many cases which need more special attention when building and providing mocked objects, this keeps saving my life in many situations. [DWIM](https://en.wikipedia.org/wiki/DWIM).

caronc commented on apprise/plugins/NotifyDBus.py:
```diff
@@ -77,19 +83,26 @@
         NOTIFY_DBUS_IMAGE_SUPPORT = True
 
-    except (ImportError, ValueError, AttributeError):
+    except (ImportError, ValueError, AttributeError) as ex:
```
I would ask that you role all of these exceptions on the module failing to load from `NotifyDBus`, `NotifyGnome`, and `NotifyMacOSX` so they are silent.  It's totally okay for these to fail.  Especially for Windows users.  They won't want to see a warning message everytime they use Apprise.

A global boolean is set that the module is disabled and this can be seen why in the `--details` for CLI users and `Apprise.details()` for API/Dev users.

amotl replied:
Now I understood your request. I think it will be fixed now?

caronc commented on apprise/plugins/NotifyMacOSX.py:
```diff
@@ -124,19 +126,17 @@ def __init__(self, sound=None, include_image=True, **kwargs):
         self.include_image = include_image
 
-        # Acquire the notify path
+        # Acquire the path to the `terminal-notifier` program.
```
You've done this in so many places where you corrected all my comments;  I would be spamming you to applaud your amazing work here.  But let me least acknowledge that i noticed this and it's fantastic! Great work. :rocket: .  I"m busted with my copy/pastes and bad updates in so many spots from other plugins.

caronc commented:
I realize this is still WIP, there are a few new things here, but otherwise this PR is strikingly similar to #687 . I imagine whichever gets merged first will provide a boatload of merge conflicts for you to resolve. No doubt you're already aware :).

I mean overall it's amazing work. Truly! You put in a great deal of effort and i really appreciate it.  just a couple things i found that stood out that i commented on.

amotl commented:
Dear Chris,

first of all, thank you for your excellent reviews.

> I realize this is still WIP.

This PR is stacked on top of #687, that's why I added the `[STACK]` prefix, and added a corresponding note to the original post: _Until #687 will be merged, [...]_. Let's focus on the other patches first, which are aiming to go straight into the master branch. Thanks a stack for your patience, and apologies for any confusion.

With kind regards,
Andreas.

amotl commented on test/test_plugin_pushbullet.py:
```diff
@@ -181,6 +181,18 @@
 )
 
 
+@pytest.fixture(scope="module", autouse=True)
+def no_throttling(module_mocker):
+    """
+    A pytest fixture which disables Apprise throttling for `NotifyPushBullet`.
+
+    This fixture, in contrast to the `no_throttling` fixture in `conftest.py`,
+    is both module-scoped and automatically applied. In this manner, it is not
+    needed to attach it to individual test cases within this module.
+    """
+    module_mocker.patch.object(NotifyPushBullet, "request_rate_per_sec", 0)
```
This implementation was the result of our conversation at https://github.com/caronc/apprise/pull/687#discussion_r992312635, thanks a stack.

> This fixture is both module-scoped and automatically applied. In this manner, it is not needed to attach it to individual test cases within this module.

Saying that, it is still possible to adjust that behavior for specific test cases. For example, if you would have a test module where one specific test case needs to run with throttling enabled, you can just slap `mocker` on your test case function, and patch the target into another direction, like
```
mocker.patch.object(NotifyPushBullet, "request_rate_per_sec", 42.42)
```

caronc replied:
The details as to what is best, I'm at the point to just let you decide. My only concern was that it be nestled in with the test file associated with the object being tested and you nailed that :rocket: Your approach (while more code) is less harsh then my monkey patching where i just fixed the request_rate_per_sec in each test.  This monkey patching still takes place in `test/helpers/rest.py` [here](https://github.com/caronc/apprise/blob/cb7f51d82a55910a1b6b4fb6adcfe9ec2835a407/test/helpers/rest.py#L476), [here](https://github.com/caronc/apprise/blob/cb7f51d82a55910a1b6b4fb6adcfe9ec2835a407/test/helpers/rest.py#L391), [here](https://github.com/caronc/apprise/blob/cb7f51d82a55910a1b6b4fb6adcfe9ec2835a407/test/helpers/rest.py#L234), and [here](https://github.com/caronc/apprise/blob/cb7f51d82a55910a1b6b4fb6adcfe9ec2835a407/test/helpers/rest.py#L119).  

Personally it doesn't bother me at all, but i can just imagine you twitching with me sharing that with you after all of the re-writing and precautionary (and proper) code you've been installing over the past few days :wink:

amotl replied:
As a followup to what we have here, I will add your suggestion from https://github.com/caronc/apprise/pull/687#issuecomment-1279338634:

>> Honestly though, if your `disable_throttle` just iterated through `common.NOTIFY_CUSTOM_MODULE_MAP` and set the values to zero for the decorator. Thoughts?
>
> I see that you are in favor of such a catch-all fixture. I will be happy to add one on behalf of #689, because I would like to use `pytest-mock` aka. `mocker` for that job. Would you accept this as a promise for the future?

amotl replied:
Did you have something like ea4d7e11b in mind? In this manner, disabling throttling works out of the box, and it seems to work well so far.

amotl replied:
Hi again. I've diverted this into a separate patch, see #701.

caronc commented on apprise/plugins/NotifyDBus.py:
```diff
@@ -77,19 +83,26 @@
         NOTIFY_DBUS_IMAGE_SUPPORT = True
 
-    except (ImportError, ValueError, AttributeError):
+    except (ImportError, ValueError, AttributeError) as ex:
```
My only comment on this thread would be to drop all the captures of the logging failures here and return to `pass` for these 3 files (DBus, Gnome, and MacOSX).  Import errors should silently fail (a global flag gets set to False for reasons later).  But we don't want Operating systems like Windows to see these errors on every single call to Apprise. :slightly_smiling_face:

amotl replied:
Thanks. I've fixed it.

amotl commented:
I think this patch may be ready now. Do you have any other suggestions?

amotl commented on test/conftest.py:
```diff
@@ -27,22 +27,17 @@
-    backup = {}
-    backup["NotifyBase"] = NotifyBase.request_rate_per_sec
-    backup["NotifyPushBullet"] = NotifyPushBullet.request_rate_per_sec
-    NotifyBase.request_rate_per_sec = 0
-    NotifyPushBullet.request_rate_per_sec = 0
-    yield
-    NotifyBase.request_rate_per_sec = backup["NotifyBase"]
-    NotifyPushBullet.request_rate_per_sec = backup["NotifyPushBullet"]
+    for notifier in NOTIFY_MODULE_MAP.values():
+        plugin = notifier["plugin"]
+        session_mocker.patch.object(plugin, "request_rate_per_sec", 0)
```
I think this makes a dream come true? Thanks for insisting to steward me into this direction. I like it.

amotl replied:
Diverted to #701.

amotl commented on test/helpers/module.py:
```diff
@@ -22,34 +22,34 @@
     reload(sys.modules['apprise.config'])
-    if module_name in sys.modules:
-        reload(sys.modules[module_name])
+    reload(sys.modules[module_name])
```
That might be dangerous because it raises a `KeyError` when the module has not been loaded yet.

caronc replied:
I think it's better they way it was honestly.  There is no test coverage in a test file, so it's not like it matters at the end of the day.

amotl replied:
I've just reverted this change.
